tests/unit_tests/tests_launch/test_launch_aws.py

```diff
@@ -32,37 +35,105 @@ def mock_create_training_job(*args, **kwargs):
     }
 
 
-def mock_boto3_client(*args, **kwargs):
-    if args[0] == "sagemaker":
-        mock_sagemaker_client = MagicMock()
-        mock_sagemaker_client.create_training_job = mock_create_training_job
-        mock_sagemaker_client.create_training_job.return_value = {
-            "TrainingJobArn": "arn:aws:sagemaker:us-east-1:123456789012:TrainingJob/test-job-1"
-        }
-        mock_sagemaker_client.stop_training_job.return_value = {
-            "TrainingJobArn": "arn:aws:sagemaker:us-east-1:123456789012:TrainingJob/test-job-1"
-        }
-        mock_sagemaker_client.describe_training_job.return_value = {
-            "TrainingJobStatus": "Completed",
-            "TrainingJobName": "test-job-1",
-        }
-        return mock_sagemaker_client
-    elif args[0] == "ecr":
-        ecr_client = MagicMock()
-        ecr_client.get_authorization_token.return_value = {
-            "authorizationData": [
-                {
-                    "proxyEndpoint": "https://123456789012.dkr.ecr.us-east-1.amazonaws.com",
-                }
-            ]
-        }
-        return ecr_client
-    elif args[0] == "sts":
+def mock_sagemaker_client():
+    mock_sagemaker_client = MagicMock()
+    mock_sagemaker_client.create_training_job = mock_create_training_job
+    mock_sagemaker_client.create_training_job.return_value = {
+        "TrainingJobArn": "arn:aws:sagemaker:us-east-1:123456789012:TrainingJob/test-job-1"
+    }
+    mock_sagemaker_client.stop_training_job.return_value = {
+        "TrainingJobArn": "arn:aws:sagemaker:us-east-1:123456789012:TrainingJob/test-job-1"
+    }
+    mock_sagemaker_client.describe_training_job.return_value = {
+        "TrainingJobStatus": "Completed",
+        "TrainingJobName": "test-job-1",
+    }
+    return mock_sagemaker_client
+
+
+def mock_ecr_client():
+    ecr_client = MagicMock()
+    ecr_client.get_authorization_token.return_value = {
+        "authorizationData": [
+            {
+                "proxyEndpoint": "https://123456789012.dkr.ecr.us-east-1.amazonaws.com",
+            }
+        ]
+    }
+    return ecr_client
+
+
+def mock_boto3_client(
+    *args,
+    **kwargs,
+):
+    sts_client = MagicMock()
+    sts_client.get_caller_identity.return_value = {"Account": "123456789012"}
+    clients = {
+        "sagemaker": mock_sagemaker_client(),
+        "ecr": mock_ecr_client(),
+        "sts": sts_client,
+    }
+    return clients[args[0]]
+
+
+def mock_boto3_client_no_instance(*args, **kwargs):
+
+    if kwargs.get("aws_access_key_id") is None:
         sts_client = MagicMock()
-        sts_client.get_caller_identity.return_value = {
-            "Account": "123456789012",
-        }
-        return sts_client
+        sts_client.get_caller_identity = MagicMock(
+            side_effect=botocore.exceptions.NoCredentialsError,
+        )
+    elif kwargs.get("aws_access_key_id") is not None:
```

Comment:
could just be `else` but this makes it explicit so probably better for readability?

Reply:
that was my take.